### PR TITLE
[#527] Moved Inventory from AttackableUnit to ObjAiBase

### PR DESCRIPTION
**Pull request by abosma**
Moved Inventory from AttackableUnit to ObjAiBase related to: https://github.com/LeagueSandbox/GameServer/issues/527

### CODE REVIEW

File: PacketDefinitions420/PacketDefinitions/S2C/BuyItemResponse.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ public BuyItemResponse(IAttackableUnit actor, IItem item, byte unk = 0x29)
             : base(PacketCmd.PKT_S2C_BUY_ITEM_ANS, actor.NetId)
         {
             Write((int)item.ItemData.ItemId);
-            Write((byte)actor.Inventory.GetItemSlot(item));
+            Write((byte)((IObjAiBase)actor).Inventory.GetItemSlot(item));
```
Comment (lzardy) [inline PacketDefinitions420/PacketDefinitions/S2C/BuyItemResponse.cs]: If you have time, it may be nice if you switched this to use LeaguePackets (where this packet is used is NotifyBuyItem and NotifyItemBought, this is where you would implement LeaguePackets' version). Take a look at NotifyMinionSpawned for some insight on how it should be structured.

Comment (abosma) [reply PacketDefinitions420/PacketDefinitions/S2C/BuyItemResponse.cs]: Sure I'll give it a go.

**Comment by abosma** [comment]
NotifyBuyItem now uses LeaguePackets, and NotifyBoughtItem is not being used anymore.

TODO: 
- Implement spell charges for items
- Find out what the bitfield is for BuyItemAns in LeaguePackets
- Maybe just remove NotifyBoughtItem

**Comment by abosma** [comment]
Added ISpellData to IItemData. This SpellData is loaded on Game.Initialize.

By doing this the packet now includes the bought item spell stacks.

**Review by FurkanS1821 — COMMENTED** [review]
I saw nothing wrong, and +1 to @microsoftv 's comment, will be approved once it's also done.

File: GameServerLib/Items/ItemData.cs [hunk 2]
```diff
@@ -62,6 +63,7 @@ public static ItemData Load(ItemManager owner, ItemContentCollectionEntry itemIn
                 Price = itemInfo.GetInt("Data", "Price"),
                 ItemGroup = itemInfo.GetString("Data", "ItemGroup"),
                 SpellName = itemInfo.GetString("Data", "SpellName"),
+                SpellData = new SpellData(game),
```
Comment (danil179) [inline GameServerLib/Items/ItemData.cs]: Please change the coupling here if you have time. From what I see, the only use in `SpellData` is for the loading:
`_game.Config.ContentManager.GetContentFileFromJson("Spells", name);`

Change this to `ContentManager` or something less general than `Game` to discourage some workarounds using game's variables.

Comment (abosma) [reply GameServerLib/Items/ItemData.cs]: I'll have to remove SpellData from ItemData if I wanted to do that since ContentManager isn't loaded in until after ItemData.

**Comment by danil179** [comment]
I think your rebase isn't correct cause the commits have different hashes.

I'll check this PRand when approved and we will see if I there are no conflicts..

File: GameServerLib/Packets/PacketHandlers/HandleSpawn.cs [hunk 3]
```diff
@@ -78,8 +78,12 @@ public override bool HandlePacket(int userId, SpawnRequest req)
...
                     foreach (var item in turret.Inventory)
                     {
-                        if (item == null) continue;
-                         _game.PacketNotifier.NotifyItemBought(turret, item as IItem);
+                        if (item == null)
+                        {
+                            continue;
+                        }
+
+                         _game.PacketNotifier.NotifyBuyItem((int)turret.NetId, turret, item as IItem);
```
Comment (danil179) [inline GameServerLib/Packets/PacketHandlers/HandleSpawn.cs]: There are 5 spaces in the indentation here. Please change it to 4.

File: GameServerLibTests/Tests/Items/TestInventoryManager.cs [hunk 4]
```diff
@@ -11,8 +13,9 @@ public class InventoryManagerTests
         [TestMethod]
         public void TestAddAndGetItem()
         {
-            var itemManager = new ItemManager();
-            itemManager.AddItems(ItemContentCollection.LoadItemsFrom("../../../Content"));
+            var game = new Game();
+            var itemManager = game.ItemManager;
```
Comment (danil179) [inline GameServerLibTests/Tests/Items/TestInventoryManager.cs]: Why to create `Game` to check the `ItemManager`? Change it to just creation of `ItemManager`.

File: GameServerLibTests/Tests/Items/TestInventoryManager.cs [hunk 5]
```diff
@@ -49,8 +52,9 @@ public void TestAddAndGetItem()
         [TestMethod]
         public void TestItemStacking()
         {
-            var itemManager = new ItemManager();
-            itemManager.AddItems(ItemContentCollection.LoadItemsFrom("../../../Content"));
+            var game = new Game();
+            var itemManager = game.ItemManager;
+            itemManager.AddItems(ItemContentCollection.LoadItemsFrom("../../../Content/LeagueSandbox-Default/Items"));
```
Comment (danil179) [inline GameServerLibTests/Tests/Items/TestInventoryManager.cs:55]: Same here.

File: GameServerLibTests/Tests/Items/TestInventoryManager.cs [hunk 6]
```diff
@@ -71,7 +75,7 @@ public void TestItemStacking()
             Assert.AreEqual(1, item2.StackCount);
 
             // Stack the second item, and make sure the second gets stacked
-            for(var i = 0; i < itemType2.MaxStack - 1; i++)
+            for (var i = 0; i < itemType2.MaxStack - 1; i++)
```
Comment (danil179) [inline GameServerLibTests/Tests/Items/TestInventoryManager.cs]: Why did you add space here?

File: GameServerLibTests/Tests/Items/TestInventoryManager.cs [hunk 7]
```diff
@@ -89,8 +93,9 @@ public void TestItemStacking()
         [TestMethod]
         public void TestSetExtraItem()
         {
-            var itemManager = new ItemManager();
-            itemManager.AddItems(ItemContentCollection.LoadItemsFrom("../../../Content"));
+            var game = new Game();
+            var itemManager = game.ItemManager;
```
Comment (danil179) [inline GameServerLibTests/Tests/Items/TestInventoryManager.cs]: Same here.

File: GameServerLibTests/Tests/Items/TestInventoryManager.cs [hunk 8]
```diff
@@ -114,8 +119,9 @@ public void TestSetExtraItem()
         [TestMethod]
         public void TestGetItemSlot()
         {
-            var itemManager = new ItemManager();
-            itemManager.AddItems(ItemContentCollection.LoadItemsFrom("../../../Content"));
+            var game = new Game();
+            var itemManager = game.ItemManager;
```
Comment (danil179) [inline GameServerLibTests/Tests/Items/TestInventoryManager.cs]: Same here.

File: GameServerLibTests/Tests/Items/TestInventoryManager.cs [hunk 9]
```diff
@@ -145,8 +151,9 @@ public void TestGetItemSlot()
         [TestMethod]
         public void TestRemoveItem()
         {
-            var itemManager = new ItemManager();
-            itemManager.AddItems(ItemContentCollection.LoadItemsFrom("../../../Content"));
+            var game = new Game();
+            var itemManager = game.ItemManager;
```
Comment (danil179) [inline GameServerLibTests/Tests/Items/TestInventoryManager.cs]: Same here.

File: GameServerLibTests/Tests/Items/TestInventoryManager.cs [hunk 10]
```diff
@@ -170,8 +177,9 @@ public void TestRemoveItem()
         [TestMethod]
         public void TestSwapItems()
         {
-            var itemManager = new ItemManager();
-            itemManager.AddItems(ItemContentCollection.LoadItemsFrom("../../../Content"));
+            var game = new Game();
+            var itemManager = game.ItemManager;
```
Comment (danil179) [inline GameServerLibTests/Tests/Items/TestInventoryManager.cs]: Same here.

File: GameServerLibTests/Tests/Items/TestInventoryManager.cs [hunk 11]
```diff
@@ -220,9 +228,10 @@ public void TestSwapItems()
         [TestMethod]
         public void TestGetAvailableItems()
         {
-            var itemManager = new ItemManager();
+            var game = new Game();
+            var itemManager = game.ItemManager;
```
Comment (danil179) [inline GameServerLibTests/Tests/Items/TestInventoryManager.cs]: Same here.

File: PacketDefinitions420/PacketNotifier.cs [hunk 12]
```diff
@@ -292,10 +292,26 @@ public void NotifyAvatarInfo(int userId, ClientInfo client)
...
+            };
+
+            var buyItemPacket = new BuyItemAns
+            {
+                Item = itemData,
+                SenderNetID = gameObject.NetId
+            };
+
+            // todo find out what bitfield does, currently unknown
+            //buyItemPacket.Bitfield = null;
```
Comment (danil179) [inline PacketDefinitions420/PacketNotifier.cs]: Remove commented code. Change `todo` -> `TODO: `

(See here: https://docs.microsoft.com/en-us/visualstudio/ide/using-the-task-list?view=vs-2019)

**Review by danil179 — CHANGES_REQUESTED** [review]
There are some style problems.

File: GameServerLibTests/Tests/Items/TestInventoryManager.cs (hunk 6 above)
Comment (abosma) [reply GameServerLibTests/Tests/Items/TestInventoryManager.cs]: Above it was done the same, so I changed it for consistency.

**Comment by danil179** [comment]
Some of the issues are unresolved.

**Comment by abosma** [comment]
Should be resolved now.

**Review by danil179 — APPROVED** [review]
Approved.
